Issue #6: Replace require_user_key_login with actual session

tomdickman opened the pull request:
The previous iteration used `require_user_key_login` to validate
user access key, this means that no real session was created and
the target URL script was imported via `require` which could
lead to errors if import scripts in target script are relative.

Remove `require_user_key_login` use and manually validate the
token before setting session user and then redirecting to the
target URL.

Commit:
Issue #6: Replace require_user_key_login with actual session

The previous iteration used require_user_key_login to validate
user access key, this means that no real session was created and
the target URL script was imported via `require` which could
lead to errors if import scripts in target script are relative.
Remove require_user_key_login use and manually validate the
token before setting session user and then redirecting to the
target URL.

dmitriim commented on classes/converter.php:
```diff
@@ -67,6 +67,10 @@ public function convert_moodle_url_to_pdf(moodle_url $url, string $key, string $
+
+        // All converters MUST implement `destroy session` following conversion, in order to prevent improper use
+        // of the session for further actions.
+        $this->destroy_session();
```
This will not guarantee that all converters will do Also if you try to shut down session will you be still able to print out the stored file? 
Maybe add this to a shutdown handler if that does make sense here? 

Or you can have convert_moodle_url_to_pdf as a final method and call another abstract method inside to actually get the file. Then you can be sure that you close session in convert_moodle_url_to_pdf?

tomdickman replied:
I have moved the conversion to PDF to a separate protected method to be overridden by converters and I have added the session termination to a `final public` method for the actual API, this way, the converters are responsible for getting the PDF content and MUST go through the proxy, and the session is always terminated, to prevent session hijacking.

dmitriim commented on classes/helper.php:
```diff
@@ -172,4 +173,42 @@ public static function get_proxy_url(moodle_url $targeturl, string $key) {
+     * @param string $key access key to use for user validation, this is required to login user and allow access of target page
+     * {@see \tool_pdfpages\helper::create_user_key}.
+     */
+    public static function login_with_key(string $key) {
```
can we have tests for this method?

tomdickman replied:
Tests added

Commit:
Issue #6: Make URL to PDF conversion method final

To ensure that proxy session is not utilised for session
hijacking, destroy the session after PDF context is
obtained through proxy pass-through to target URL for
PDF generation.
Move the PDF generation into a protected method in
converters and make abstract class public method final
to force proxy pass-through and session termination.

Commit:
Issue #6: Add test coverage for access key login

dmitriim commented on classes/converter.php:
```diff
@@ -25,6 +25,7 @@
 
 namespace tool_pdfpages;
 
+use Matrix\Exception;
```
what is this for?

tomdickman replied:
Nothing, I accidentally imported it instead of `\Exception`, removing

dmitriim commented on classes/converter.php:
```diff
@@ -44,6 +45,26 @@ abstract class converter {
+     *
+     * @return string raw PDF content of URL.
+     */
+    protected function generate_pdf_content(moodle_url $proxyurl, string $filename = '', array $options = [],
```
this one should be abstract to make sure that children implement it.

dmitriim replied:
also should it be named as generate_pdf_file? as that what the method does.

dmitriim replied:
ah scratch it. It does generate a content and return a string.

tomdickman replied:
Abstracted the method

dmitriim commented on classes/converter.php:
```diff
@@ -53,20 +74,29 @@ abstract class converter {
+            throw new \moodle_exception('error:urltopdf', 'tool_pdfpages', '', null, $exception->getMessage());
+        } finally {
+            // Make sure the access key token session cannot be used for any other requests, prevent session hijacking.
+            \core\session\manager::terminate_current();
```
I'm still not convinced that we will reach it in all scenarios. Is there any way to add unit tests? Behat tests?

tomdickman replied:
There can be no behat tests as this is happening in a separate process, finally is ALWAYS carried out as per PHP https://www.php.net/manual/en/language.exceptions.php `All finally blocks it encounters along the way will be executed.`

dmitriim replied:
can we prove that in tests?

dmitriim replied:
another thing to test is that we still can access returned file after the session is terminated. Will we be able to read the content of the file?

tomdickman replied:
I have added session termination test for the abstract converter class

Commit:
Issue #6: Make PDF generation method abstract

Force conversion drivers to implement their own
PDF generation logic

Commit:
Issue #6: Add converter session termination test

Commit:
Issue #6: Add missing options validation to wkthmltopdf generation

Commit:
Issue #6: chore - remove unused class import
